src/mysql-util/src/schemas.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mysql_async::prelude::Queryable;
+use mysql_async::Conn;
+use tracing::warn;
+
+use crate::desc::{MySqlColumnDesc, MySqlDataType, MySqlTableDesc};
+use crate::MySqlError;
+
+/// Retrieve the tables and column descriptions for tables in the given schemas.
+pub async fn schema_info(
+    conn: &mut Conn,
+    schemas: Vec<&str>,
+) -> Result<Vec<MySqlTableDesc>, MySqlError> {
+    // Get all tables of type 'Base Table' in schema
+    let mut tables = conn
+        .exec_map(
+            "SELECT table_name, table_schema
+             FROM information_schema.tables
+             WHERE table_type = 'BASE TABLE'
+             AND table_schema IN (?)",
+            schemas,
+            |(table_name, table_schema)| MySqlTableDesc {
+                schema_name: table_schema,
+                name: table_name,
+                columns: Vec::new(),
+            },
+        )
+        .await?;
+
+    for table in &mut tables {
```

Comment:
I realize you are nowhere near done with this but I don't feel good about firing a warn on something that should be an error--you can relatively easily handle the error case more directly via something like:

```rust
let columns: Result<Vec<_>, _> = conn
    .exec_map(
        "SELECT column_name, data_type, is_nullable, ordinal_position
        FROM information_schema.columns
        WHERE table_name = ? AND table_schema = ?",
        (&table.name, &table.schema_name),
        |(column_name, data_type, is_nullable, ordinal_position): (
            String,
            String,
            String,
            u16,
        )| {
            let data_type: String = data_type;
            let data_type = match data_type.as_str() {
                "int" => MySqlDataType::Int,
                "varchar" => MySqlDataType::Varchar(usize::MAX),
                "char" => MySqlDataType::Varchar(usize::MAX),
                _ => {
                    return Err(anyhow::anyhow!(
                        "Skipping unsupported data type: {}",
                        data_type
                    ));
                }
            };
            Ok(MySqlColumnDesc {
                name: column_name,
                data_type,
                nullable: is_nullable == "YES",
                ordinal_position,
            })
        },
    )
    .await?
    .into_iter()
    .collect();
table.columns.append(&mut columns?);
```

The "hack" here is declaring that you want to collect into a `Result` type and that will move the inner errors of the `Vec` outside of it. I'm sure there's some way to code golf this into a form of that doesn't require the allocation that `collect` suggests it will do--maybe someone else knows that idiom offhand.